File: android/app/src/main/java/com/airbridge/app/mirror/MirrorSession.kt
package com.airbridge.app.mirror

import com.airbridge.app.core.AirBridgeLog
import com.airbridge.app.core.interfaces.IMirrorSession
import com.airbridge.app.core.interfaces.InputEventArgs
import com.airbridge.app.core.interfaces.InputEventType
import com.airbridge.app.core.interfaces.ITransferSession
import com.airbridge.app.core.interfaces.MirrorMode
import com.airbridge.app.core.interfaces.MirrorState
import com.airbridge.app.transport.interfaces.IMessageChannel
import com.airbridge.app.transport.protocol.MessageType
import com.airbridge.app.transport.protocol.ProtocolMessage
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

/**
 * Android-side implementation of [IMirrorSession].
 *
 * Responsibilities:
 * - Waits for a [MessageType.MIRROR_START] request from the Windows host.
 * - Starts screen capture and streams H.264 frames (MediaProjection wiring in Iteration 6).
 * - On [MessageType.MIRROR_STOP], tears down capture and stops.
 * - Receives [MessageType.INPUT_EVENT] messages and injects them via [InputInjector].
 * - **Passes through** file-transfer messages to an optional [ITransferSession] receiver
 *   so drag-and-drop file transfers work while a mirror session is active.
 *
 * @param sessionId           Unique identifier for this session.
 * @param channel             Transport channel to the Windows host.
 * @param captureSession      Optional running [ScreenCaptureSession]; null in headless/test mode.
 * @param width               Capture width in pixels.
 * @param height              Capture height in pixels.
 * @param fps                 Target frame rate.
 * @param codec               Codec string (default "H264").
 * @param inputInjector       Optional [InputInjector] for relaying input events from Windows.
 *                            Pass null for view-only mode.
 * @param fileTransferReceiver Optional transfer session that handles incoming file-transfer
 *                             messages. Pass null to ignore file-transfer traffic.
 * @param coroutineScope      Scope for background coroutines. Inject a test scope for
 *                            deterministic unit testing.
 */
class MirrorSession(
    override val sessionId: String,
    private val channel: IMessageChannel,
    private val captureSession: ScreenCaptureSession? = null,
    private val width: Int = 0,
    private val height: Int = 0,
    private val fps: Int = 30,
    private val codec: String = "H264",
    private val inputInjector: InputInjector? = null,
    private val fileTransferReceiver: ITransferSession? = null,
    private val coroutineScope: CoroutineScope = CoroutineScope(SupervisorJob() + Dispatchers.IO),
) : IMirrorSession {

    // ── State ──────────────────────────────────────────────────────────────

    private val _stateFlow = MutableStateFlow(MirrorState.CONNECTING)
    override val stateFlow: Flow<MirrorState> = _stateFlow.asStateFlow()
    override val mode: MirrorMode = MirrorMode.PHONE_WINDOW

    private var receiveJob: Job? = null
    private var inputReceiveJob: Job? = null

    // ── IMirrorSession lifecycle ───────────────────────────────────────────

    /**
     * Starts the mirror session: sends [MirrorStartMessage] to Windows, begins
     * streaming frames from [captureSession] (if provided), and launches the
     * input relay and file-transfer receive loops.
     */
    override suspend fun start() {
        check(_stateFlow.value == MirrorState.CONNECTING) {
            "Cannot start a session in state ${_stateFlow.value}."
        }
        AirBridgeLog.info("[Mirror:$sessionId] CONNECTING — starting session (${width}x${height} @${fps}fps codec=$codec)")

        // Announce the mirror session to the Windows host
        if (captureSession != null && width > 0 && height > 0) {
            val startMsg = MirrorStartMessage(
                mode      = MirrorSessionMode.PHONE_WINDOW,
                codec     = if (codec == "H265") MirrorCodec.H265 else MirrorCodec.H264,
                width     = width,
                height    = height,
                fps       = fps.toByte(),
                sessionId = sessionId
            )
            channel.send(
                ProtocolMessage(
                    type    = MessageType.MIRROR_START,
                    payload = startMsg.toBytes()
                )
            )
        }

        // Update injector screen dimensions for coordinate de-normalisation
        inputInjector?.let {
            it.screenWidth  = width
            it.screenHeight = height
        }

        // When captureSession is present this session is the *sender* — DeviceConnectionService
        // already owns the sole read loop on channel.incomingMessages.  Launching additional
        // collect() calls here would compete for bytes on the DataInputStream, corrupting the
        // framing and breaking the keepalive.  INPUT_EVENT and MIRROR_STOP are handled via
        // DeviceConnectionService.addMessageHandler — see createMessageHandler() below.
        if (captureSession == null) {
            // Receiver mode (no capture source): handle incoming control messages directly.
            if (inputInjector != null) {
                inputReceiveJob = coroutineScope.launch {
                    try { runInputReceiveLoop() } catch (_: CancellationException) { }
                }
            }

            receiveJob = coroutineScope.launch {
                try {
                    runReceiveLoop()
                } catch (e: CancellationException) {
                    AirBridgeLog.info("[Mirror:$sessionId] Receive loop cancelled → STOPPED")
                    _stateFlow.value = MirrorState.STOPPED
                    throw e
                } catch (e: Exception) {
                    AirBridgeLog.error("[Mirror:$sessionId] Receive loop exception → ERROR", e)
                    _stateFlow.value = MirrorState.ERROR
                }
            }
        }

        // Stream frames from capture session if available
        if (captureSession != null) {
            AirBridgeLog.info("[Mirror:$sessionId] CONNECTING → ACTIVE; streaming frames from captureSession")
            _stateFlow.value = MirrorState.ACTIVE
            captureSession.frames.collect { frame ->
                try {
                    channel.send(ProtocolMessage(MessageType.MIRROR_FRAME, frame.toBytes()))
                } catch (e: Exception) {
                    AirBridgeLog.error("[Mirror:$sessionId] Frame send failed → ERROR", e)
                    _stateFlow.value = MirrorState.ERROR
                    return@collect
                }
            }
            AirBridgeLog.info("[Mirror:$sessionId] captureSession.frames completed")
        }
    }

    /**
     * Stops the session gracefully: sends [MirrorStopMessage] and cancels all coroutines.
     */
    override suspend fun stop() {
        try {
            channel.send(ProtocolMessage(MessageType.MIRROR_STOP, ByteArray(0)))
        } catch (_: Exception) {
            // Best-effort; channel may already be closed.
        }
        inputReceiveJob?.cancel()
        receiveJob?.cancel()
        _stateFlow.value = MirrorState.STOPPED
    }

    /**
     * Injects an [InputEventArgs] locally via [InputInjector].
     * No-op when no injector was supplied (view-only mode).
     */
    override suspend fun sendInput(event: InputEventArgs) {
        inputInjector?.inject(event)
    }

    // ── Sender-mode message handler ───────────────────────────────────────

    /**
     * Returns a handler delegate that routes [INPUT_EVENT] and [MIRROR_STOP] messages
     * into this session when it is running in sender mode (captureSession != null).
     *
     * Register this with [com.airbridge.app.core.DeviceConnectionService.addMessageHandler]
     * immediately after [start] is called so Windows-originated control messages reach
     * the session without competing with the DeviceConnectionService read loop.
     *
     * @param onStop Callback invoked when [MIRROR_STOP] is received from Windows.
     *               Typically calls [PhoneCaptureService.stopCapture].
     */
    fun createMessageHandler(onStop: () -> Unit): suspend (ProtocolMessage) -> Unit = { msg ->
        when (msg.type) {
            MessageType.MIRROR_START -> {
                // Windows sends MirrorStart after its decode pipeline is ready to signal
                // "I'm set up — please send me a fresh IDR keyframe now."
                AirBridgeLog.info("[Mirror:$sessionId] MIRROR_START from Windows → requesting IDR keyframe")
                captureSession?.requestKeyFrame()
            }
            MessageType.INPUT_EVENT -> {
                try {
                    val inputMsg = InputEventMessage.fromBytes(msg.payload)
                    val evtType = when (inputMsg.eventKind) {
                        InputEventKind.TOUCH -> InputEventType.TOUCH
                        InputEventKind.KEY   -> InputEventType.KEY
                        InputEventKind.MOUSE -> InputEventType.MOUSE
                    }
                    sendInput(InputEventArgs(
                        type        = evtType,
                        normalizedX = inputMsg.normalizedX,
                        normalizedY = inputMsg.normalizedY,
                        keycode     = inputMsg.keycode,
                        metaState   = inputMsg.metaState
                    ))
                } catch (_: Exception) {
                    // Malformed message — drop
                }
            }
            MessageType.MIRROR_STOP -> {
                AirBridgeLog.info("[Mirror:$sessionId] MIRROR_STOP from Windows → stopping capture")
                onStop()
            }
            else -> { /* ignore */ }
        }
    }

    // ── Input relay receive loop ───────────────────────────────────────────

    /**
     * Collects [MessageType.INPUT_EVENT] messages from the channel and dispatches
     * them to [inputInjector]. Unknown or malformed messages are silently skipped.
     */
    private suspend fun runInputReceiveLoop() {
        channel.incomingMessages.collect { msg ->
            if (msg.type != MessageType.INPUT_EVENT) return@collect
            try {
                val inputMsg = InputEventMessage.fromBytes(msg.payload)
                val evtType = when (inputMsg.eventKind) {
                    InputEventKind.TOUCH -> InputEventType.TOUCH
                    InputEventKind.KEY   -> InputEventType.KEY
                    InputEventKind.MOUSE -> InputEventType.MOUSE
                }
                val args = InputEventArgs(
                    type        = evtType,
                    normalizedX = inputMsg.normalizedX,
                    normalizedY = inputMsg.normalizedY,
                    keycode     = inputMsg.keycode,
                    metaState   = inputMsg.metaState
                )
                inputInjector?.inject(args)
            } catch (_: Exception) {
                // Malformed message — skip
            }
        }
    }

    // ── Main receive loop ──────────────────────────────────────────────────

    /**
     * Processes mirror-protocol messages and forwards file-transfer messages
     * to [fileTransferReceiver].
     */
    private suspend fun runReceiveLoop() {
        channel.incomingMessages.collect { msg ->
            when (msg.type) {
                MessageType.MIRROR_START -> {
                    AirBridgeLog.info("[Mirror:$sessionId] MIRROR_START received → ACTIVE")
                    _stateFlow.value = MirrorState.ACTIVE
                }

                MessageType.MIRROR_STOP -> {
                    AirBridgeLog.info("[Mirror:$sessionId] MIRROR_STOP received → STOPPED")
                    _stateFlow.value = MirrorState.STOPPED
                    receiveJob?.cancel()
                }

                // File-transfer pass-through: Windows sends these when the user drops a
                // file onto the MirrorWindow. Forward to the optional transfer receiver.
                MessageType.FILE_TRANSFER_START,
                MessageType.FILE_CHUNK,
                MessageType.FILE_TRANSFER_END,
                MessageType.FILE_TRANSFER_ACK -> {
                    forwardToTransferReceiver(msg)
                }

                // Input events are handled by runInputReceiveLoop — ignore here
                MessageType.INPUT_EVENT -> { /* handled separately */ }

                else -> { /* no-op for forward-compatibility */ }
            }
        }
    }

    private fun forwardToTransferReceiver(msg: ProtocolMessage) {
        fileTransferReceiver ?: return
        // Payload bytes are available for the service layer to route into a TransferSession.
    }

    // ── Cleanup ────────────────────────────────────────────────────────────

    /** Cancels all background coroutines and releases resources. */
    fun dispose() {
        inputReceiveJob?.cancel()
        receiveJob?.cancel()
        coroutineScope.cancel()
    }
}

/**
 * High-level mirror service — creates [MirrorSession] instances and tracks active sessions.
 */
class MirrorService @Inject constructor(
    private val inputInjector: InputInjector
) : com.airbridge.app.mirror.interfaces.IMirrorService {

    private val _activeSessions = mutableListOf<IMirrorSession>()

    override suspend fun startMirror(
        remoteDevice: com.airbridge.app.core.models.DeviceInfo,
        mode: MirrorMode
    ): IMirrorSession {
        // For TABLET_DISPLAY the session lifecycle is owned by TabletDisplayActivity (it
        // creates TabletDisplaySession once the SurfaceView surface is ready). Return a
        // lightweight stub so callers can observe state transitions without crashing.
        // For PHONE_WINDOW a channel + capture session are required; use startMirrorWithChannel().
        if (mode == MirrorMode.TABLET_DISPLAY) {
            val stub = TabletDisplayStubSession()
            _activeSessions.add(stub)
            return stub
        }
        throw UnsupportedOperationException(
            "Use startMirrorWithChannel() — a message channel and capture session are required."
        )
    }

    /**
     * Creates a [MirrorSession] wired to [channel] and [captureSession].
     */
    fun startMirrorWithChannel(
        sessionId: String,
        channel: IMessageChannel,
        captureSession: ScreenCaptureSession,
        width: Int,
        height: Int,
        fps: Int,
        codec: String = "H264"
    ): MirrorSession {
        val session = MirrorSession(
            sessionId      = sessionId,
            channel        = channel,
            captureSession = captureSession,
            width          = width,
            height         = height,
            fps            = fps,
            codec          = codec,
            inputInjector  = inputInjector
        )
        _activeSessions.add(session)
        return session
    }

    override fun getActiveSessions(): List<IMirrorSession> = _activeSessions.toList()
}

/**
 * Lightweight stub session returned by [MirrorService.startMirror] for
 * [MirrorMode.TABLET_DISPLAY]. The real session is created inside
 * [com.airbridge.app.display.TabletDisplayActivity] once the SurfaceView
 * surface is available. This stub lets [com.airbridge.app.ui.viewmodels.MirrorViewModel]
 * observe state without holding a reference to the actual MediaCodec session.
 */
private class TabletDisplayStubSession : IMirrorSession {
    override val sessionId: String = "tablet-display-stub-${System.currentTimeMillis()}"
    override val mode: MirrorMode  = MirrorMode.TABLET_DISPLAY

    private val _stateFlow = MutableStateFlow(MirrorState.ACTIVE)
    override val stateFlow: Flow<MirrorState> = _stateFlow.asStateFlow()

    override suspend fun start()  { /* no-op: real session lives in TabletDisplayActivity */ }
    override suspend fun stop()   { _stateFlow.value = MirrorState.STOPPED }
    override suspend fun sendInput(event: InputEventArgs) { /* no-op */ }
}

File: windows/AirBridge.Mirror/MirrorSession.cs
using AirBridge.Core.Interfaces;
using AirBridge.Transport.Interfaces;
using AirBridge.Transport.Protocol;

namespace AirBridge.Mirror;

/// <summary>
/// Windows-side implementation of <see cref="IMirrorSession"/>.
/// <para>
/// Responsibilities:
/// <list type="bullet">
///   <item>Sends <see cref="MessageType.MirrorStart"/> to the Android device and
///         opens a floating window via the injected <see cref="IMirrorWindowHost"/>.</item>
///   <item>Feeds <see cref="MessageType.MirrorFrame"/> payloads to the
///         <see cref="IMirrorDecoder"/> and presents them in the window.</item>
///   <item>On <see cref="MessageType.MirrorStop"/>, closes the window and stops.</item>
///   <item>Captures pointer/keyboard events from the window and relays them to
///         Android as <see cref="InputEventMessage"/> (0x30) messages.</item>
///   <item>When files are dropped onto the window, calls <see cref="SendFileAsync"/>
///         which delegates to the injected <see cref="ITransferEngine"/>.</item>
///   <item>Passes through <see cref="MessageType.FileTransferAck"/> /
///         <c>FileTransferEnd</c> messages received from Android.</item>
/// </list>
/// </para>
/// </summary>
public sealed class MirrorSession : IMirrorSession
{
    // ── Dependencies ───────────────────────────────────────────────────────

    private readonly IMessageChannel                          _channel;
    private readonly Func<IMirrorDecoder>?                    _decoderFactory;
    private readonly Func<IMirrorDecoder, IMirrorWindowHost>? _windowFactory;
    private readonly ITransferEngine?                         _transferEngine;
    private readonly IProgress<long>?                         _transferProgress;
    private readonly bool                                     _androidInitiated;
    private readonly int                                      _width;
    private readonly int                                      _height;

    // ── Runtime state ──────────────────────────────────────────────────────

    private MirrorState         _state = MirrorState.Connecting;
    private CancellationTokenSource? _cts;
    private IMirrorWindowHost?  _window;
    private IMirrorDecoder?     _decoder;
    private readonly object     _stateLock = new();

    // ── Inbound message queue ─────────────────────────────────────────────
    // MirrorSession must NOT call _channel.ReceiveAsync() directly because
    // DeviceConnectionService.MonitorSessionAsync is already the single reader
    // on the channel's SslStream (concurrent reads throw NotSupportedException).
    // Instead, the caller registers CreateMessageHandler() with DeviceConnectionService
    // so dispatched messages are written here and the session reads only from this queue.
    private readonly System.Threading.Channels.Channel<ProtocolMessage> _inboundQueue =
        System.Threading.Channels.Channel.CreateUnbounded<ProtocolMessage>(
            new System.Threading.Channels.UnboundedChannelOptions { SingleReader = true });

    // ── Input send queue ───────────────────────────────────────────────────
    // Bounded channel so SendInputAsync is non-blocking and fire-and-forget safe.
    private readonly System.Threading.Channels.Channel<InputEventArgs> _inputQueue =
        System.Threading.Channels.Channel.CreateBounded<InputEventArgs>(
            new System.Threading.Channels.BoundedChannelOptions(64)
            {
                FullMode     = System.Threading.Channels.BoundedChannelFullMode.DropOldest,
                SingleReader = true
            });

    // ── IMirrorSession ─────────────────────────────────────────────────────

    /// <inheritdoc/>
    public string SessionId { get; }

    /// <inheritdoc/>
    public MirrorMode Mode { get; }

    /// <inheritdoc/>
    public MirrorState State
    {
        get { lock (_stateLock) return _state; }
        private set
        {
            lock (_stateLock) _state = value;
            StateChanged?.Invoke(this, value);
        }
    }

    /// <inheritdoc/>
    public event EventHandler<MirrorState>? StateChanged;

    // ── Constructor ────────────────────────────────────────────────────────

    /// <summary>
    /// Creates a new <see cref="MirrorSession"/>.
    /// </summary>
    /// <param name="sessionId">Unique identifier for this session.</param>
    /// <param name="channel">The transport channel to the paired Android device.</param>
    /// <param name="decoderFactory">Factory for <see cref="IMirrorDecoder"/>; null = headless.</param>
    /// <param name="windowFactory">Factory for <see cref="IMirrorWindowHost"/>; null = headless.</param>
    /// <param name="transferEngine">Engine for drag-and-drop file transfer; null = disabled.</param>
    /// <param name="transferProgress">Optional progress reporter (bytes transferred).</param>
    public MirrorSession(
        string sessionId,
        IMessageChannel channel,
        Func<IMirrorDecoder>? decoderFactory = null,
        Func<IMirrorDecoder, IMirrorWindowHost>? windowFactory = null,
        ITransferEngine? transferEngine = null,
        IProgress<long>? transferProgress = null,
        bool androidInitiated = false,
        int width = 0,
        int height = 0)
    {
        SessionId         = sessionId ?? throw new ArgumentNullException(nameof(sessionId));
        _channel          = channel   ?? throw new ArgumentNullException(nameof(channel));
        Mode              = MirrorMode.PhoneWindow;
        _decoderFactory   = decoderFactory;
        _windowFactory    = windowFactory;
        _transferEngine   = transferEngine;
        _transferProgress = transferProgress;
        _androidInitiated = androidInitiated;
        _width            = width  > 0 ? width  : 1080;  // fallback for Windows-initiated
        _height           = height > 0 ? height : 2340;
    }

    // ── IMirrorSession lifecycle ───────────────────────────────────────────

    /// <inheritdoc/>
    public async Task StartAsync(CancellationToken cancellationToken = default)
    {
        lock (_stateLock)
        {
            if (_state != MirrorState.Connecting)
                throw new InvalidOperationException($"Cannot start a session in state {_state}.");
        }

        _cts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
        AirBridge.Core.AppLog.Info($"[Mirror:{SessionId}] Connecting → starting receive + input-send loops");

        try
        {
            // Run the message pump and input-send loop concurrently
            var receiveTask   = HandleMirrorStartAsync(_cts.Token);
            var inputSendTask = RunInputSendLoopAsync(_cts.Token);
            await Task.WhenAll(receiveTask, inputSendTask).ConfigureAwait(false);
        }
        catch (OperationCanceledException)
        {
            if (State is not MirrorState.Stopped and not MirrorState.Error)
            {
                AirBridge.Core.AppLog.Info($"[Mirror:{SessionId}] Cancelled → Stopped");
                State = MirrorState.Stopped;
            }
        }
        catch (Exception ex)
        {
            AirBridge.Core.AppLog.Error($"[Mirror:{SessionId}] Unhandled exception → Error", ex);
            State = MirrorState.Error;
            throw;
        }
        finally
        {
            _inputQueue.Writer.TryComplete();
            CleanUp();
        }
    }

    /// <inheritdoc/>
    public async Task StopAsync()
    {
        _inputQueue.Writer.TryComplete();
        _cts?.Cancel();

        try
        {
            await _channel.SendAsync(
                new ProtocolMessage(MessageType.MirrorStop, Array.Empty<byte>()))
                .ConfigureAwait(false);
        }
        catch
        {
            // Best-effort; channel may already be closed
        }

        // Null out _window before closing so CleanUp() in the StartAsync finally block
        // cannot attempt a second close on an already-closed WinUI 3 window (which crashes).
        var win = _window;
        _window = null;
        win?.Close();
        State = MirrorState.Stopped;
    }

    /// <summary>
    /// Enqueues an input event for forwarding to the connected Android device.
    /// Non-blocking: if the queue is full the oldest pending event is dropped.
    /// </summary>
    public Task SendInputAsync(
        InputEventArgs inputEvent,
        CancellationToken cancellationToken = default)
    {
        if (inputEvent is null) return Task.CompletedTask;
        _inputQueue.Writer.TryWrite(inputEvent);
        return Task.CompletedTask;
    }

    // ── Inbound message delivery ──────────────────────────────────────────

    /// <summary>
    /// Returns a handler delegate that delivers an incoming <see cref="ProtocolMessage"/>
    /// into this session's internal queue.  Register this with
    /// <c>DeviceConnectionService.AddMessageHandler</c> so that
    /// <see cref="AirBridge.App.Services.DeviceConnectionService"/> — the sole
    /// <see cref="IMessageChannel"/> reader — routes mirror-related messages here instead
    /// of reading the channel a second time (which would throw
    /// <see cref="NotSupportedException"/> on <see cref="System.Net.Security.SslStream"/>).
    /// </summary>
    public Func<ProtocolMessage, Task> CreateMessageHandler() =>
        msg =>
        {
            _inboundQueue.Writer.TryWrite(msg);
            return Task.CompletedTask;
        };

    // ── Core receive / message pump ────────────────────────────────────────

    private async Task HandleMirrorStartAsync(CancellationToken ct)
    {
        // Announce the mirror session to Android — only when Windows initiated.
        // When Android initiated (user tapped mirror on phone), Android already knows it's
        // mirroring so we must NOT echo MirrorStart back or it will confuse the protocol.
        if (!_androidInitiated)
        {
            AirBridge.Core.AppLog.Info($"[Mirror:{SessionId}] Connecting → sending MirrorStart to Android");
            await _channel.SendAsync(
                new ProtocolMessage(MessageType.MirrorStart, Array.Empty<byte>()), ct)
                .ConfigureAwait(false);
        }
        else
        {
            AirBridge.Core.AppLog.Info($"[Mirror:{SessionId}] Connecting → Android-initiated, skipping MirrorStart send");
        }

        // Create decoder (headless-safe: decoder can exist without a window)
        if (_decoderFactory is not null)
        {
            _decoder = _decoderFactory();

            // Initialize the decoder with the stream resolution before any frames arrive.
            // _width/_height come from the MirrorStart payload (Android-initiated) or
            // fall back to 1080×2340 (Windows-initiated, Android will match its capture size).
            await _decoder.InitializeAsync(_width, _height).ConfigureAwait(false);

            if (_windowFactory is not null)
            {
                _window = _windowFactory(_decoder);

                // Wire input relay: window raises events → enqueue for send loop
                _window.InputEventRaised += (_, evt) => _inputQueue.Writer.TryWrite(evt);

                // Wire drag-and-drop callback
                _window.OnFilesDropped = async files =>
                {
                    foreach (var file in files)
                        await SendFileAsync(file, _cts?.Token ?? default).ConfigureAwait(false);
                };

                // Attach the decoder's MediaStreamSource to the window's video surface
                // so WMF begins pulling NAL units and decoded frames appear on screen.
                _window.AttachDecoder(_decoder);

                // Open at phone resolution (scaled to fit screen), always-on-top.
                _window.Open(_width, _height);
            }
        }

        AirBridge.Core.AppLog.Info($"[Mirror:{SessionId}] Connecting → Active (decoder={(  _decoder is not null ? "yes" : "headless")}, window={(_window is not null ? "yes" : "none")})");
        State = MirrorState.Active;

        // Request a fresh IDR keyframe from Android.  The first IDR likely arrived before
        // our decode pipeline was fully set up (~1s of window/decoder init) and was dropped.
        // Sending MirrorStart signals Android to force a sync frame immediately.
        if (_androidInitiated)
        {
            try
            {
                await _channel.SendAsync(
                    new ProtocolMessage(MessageType.MirrorStart, Array.Empty<byte>()), ct)
                    .ConfigureAwait(false);
                AirBridge.Core.AppLog.Info($"[Mirror:{SessionId}] Sent MirrorStart → requested IDR from Android");
            }
            catch { /* best-effort */ }
        }

        // Message pump — reads from the internal queue, NOT from _channel.ReceiveAsync().
        // DeviceConnectionService is the sole channel reader; it dispatches messages here
        // via the handler returned by CreateMessageHandler().
        await foreach (var msg in _inboundQueue.Reader.ReadAllAsync(ct).ConfigureAwait(false))
        {
            switch (msg.Type)
            {
                case MessageType.MirrorStart:
                    // Android echoes MirrorStart back when it receives ours (Windows-initiated).
                    // Ignore it here — DeviceConnectionService raises AndroidMirrorStartRequested
                    // separately; we must not act on it a second time inside the running session.
                    AirBridge.Core.AppLog.Info($"[Mirror:{SessionId}] Ignoring MirrorStart echo from Android (already active)");
                    break;

                case MessageType.MirrorFrame:
                    AirBridge.Core.AppLog.Debug($"[Mirror:{SessionId}] MirrorFrame payload={msg.Payload.Length}B");
                    if (_decoder is not null)
                    {
                        var frame = MirrorFrameMessage.FromBytes(msg.Payload);
                        await _decoder.PushFrameAsync(frame.NalData, frame.IsKeyFrame, frame.PresentationTimestampUs, ct).ConfigureAwait(false);
                    }
                    break;

                case MessageType.MirrorStop:
                    AirBridge.Core.AppLog.Info($"[Mirror:{SessionId}] Active → Stopped (MirrorStop received)");
                    _window?.Close();
                    State = MirrorState.Stopped;
                    _cts?.Cancel(); // unblock RunInputSendLoopAsync
                    return;

                // ACKs from the Android transfer receiver — forwarded, not consumed here
                case MessageType.FileTransferAck:
                case MessageType.FileTransferEnd:
                    break;
            }
        }

        AirBridge.Core.AppLog.Info($"[Mirror:{SessionId}] Message pump exited → Stopped");
        _window?.Close();
        State = MirrorState.Stopped;
        _cts?.Cancel(); // unblock RunInputSendLoopAsync so Task.WhenAll can complete
    }

    // ── Input send loop ────────────────────────────────────────────────────

    private async Task RunInputSendLoopAsync(CancellationToken ct)
    {
        await foreach (var evt in _inputQueue.Reader.ReadAllAsync(ct).ConfigureAwait(false))
        {
            var kind = evt.Type switch
            {
                InputEventType.Touch => InputEventKind.Touch,
                InputEventType.Key   => InputEventKind.Key,
                _                    => InputEventKind.Mouse
            };

            var msg = new InputEventMessage(
                SessionId:   SessionId,
                EventKind:   kind,
                NormalizedX: evt.NormalizedX,
                NormalizedY: evt.NormalizedY,
                Keycode:     evt.Keycode,
                MetaState:   evt.MetaState);

            try
            {
                await _channel.SendAsync(
                    new ProtocolMessage(MessageType.InputEvent, msg.ToBytes()), ct)
                    .ConfigureAwait(false);
            }
            catch (OperationCanceledException) { break; }
            catch { break; /* Channel dead — stop the loop */ }
        }
    }

    // ── File transfer ──────────────────────────────────────────────────────

    /// <summary>
    /// Sends a dropped file to the connected Android device using the injected
    /// <see cref="ITransferEngine"/>. No-op when no engine is injected.
    /// </summary>
    public async Task SendFileAsync(IDroppedFile file, CancellationToken ct)
    {
        if (_transferEngine is null) return;

        await using var adapter = new ChannelNetworkStreamAdapter(_channel, ct);
        var result = await _transferEngine.SendFileAsync(
            file.Path, adapter, _transferProgress, ct).ConfigureAwait(false);
        _ = result;
    }

    // ── Cleanup ────────────────────────────────────────────────────────────

    private void CleanUp()
    {
        // Complete the inbound queue so ReadAllAsync exits if it is still running.
        _inboundQueue.Writer.TryComplete();
        try { _window?.Close(); }   catch { }
        try { _decoder?.Dispose(); } catch { }
        _decoder = null;
        _window  = null;
    }

    /// <inheritdoc/>
    public void Dispose()
    {
        _inputQueue.Writer.TryComplete();
        _cts?.Cancel();
        _cts?.Dispose();
        CleanUp();
    }
}
